packages/styled-components/src/models/ComponentStyle.ts

```diff
@@ -30,6 +31,7 @@ export default class ComponentStyle {
     this.componentId = componentId;
     this.baseHash = phash(SEED, componentId);
     this.baseStyle = baseStyle;
+    this.buffer = [];
```

Comment:
Seems like the failing test snapshots could be related to adding a buffer for CSSOM writes?

Reply:
@quantizor looks like the failing tests were due to parent styles not flushing, as the implementation based on #3821 never called the equivalent to `this.baseStyle.flushStyles`.

I've refactored the implementation to merge the buffer and return it with `generateStyles` (which replaces `generateAndInjectStyles`), ensuring there's no behavioural differences on React 17 and older, or during `ServerStyleSheet` usage.

While React 18 in the browser is able to make full use of `useInsertionEffect` to do the flushing, letting React coordinate the timing 😄